### PR TITLE
[WIP] Implemented flexible mechanism for supporting custom types.

### PR DESCRIPTION
**Pull request by sdanyliv**
Originally i have encountered problem when tried to use custom range types with `tstzrange`. Threre is only one type which implements timestamp range `NpgsqlRange<DateTime>` in Npgsql and there is no way to tell current engine that this particular convertation should be `timestam with range`. Analysis showed that we have to improve mapping logic.

This PR also help in the following case: #1219
- [x] Fix tests

### CODE REVIEW

**Comment by jogibear9988** [comment]
As this changes a public API:

public virtual void SetParameter(IDbDataParameter parameter, string name, DataType dataType, object value)		 		

public virtual void SetParameter(IDbDataParameter parameter, string name, DbDataType dataType, object value)

also major should be increased when this is merged

**Comment by MaceWindu** [comment]
We decided to include this pr in 2.x despite minor breaking changes for provider implementors as it contains very important changes we need for multiple features.

File: Source/LinqToDB/DataProvider/PostgreSQL/PostgreSQLDataProvider.cs [hunk 1]
```diff
@@ -441,7 +447,20 @@ protected override void SetParameterType(IDbDataParameter parameter, DataType da
...
 				case DataType.BinaryJson : if (_setJsonb     != null) _setJsonb(parameter);     else base.SetParameterType(parameter, dataType); break;
-				default                  : base.SetParameterType(parameter, dataType); break;
+				default :     
+				{
+					if (!string.IsNullOrEmpty(dataType.DbType))
+					{
+						var nativeType = GetNativeType(dataType.DbType);
+						if (nativeType != null)
+						{
+							_setNativeParameterType(parameter, nativeType);
```
Comment (MaceWindu) [inline Source/LinqToDB/DataProvider/PostgreSQL/PostgreSQLDataProvider.cs]: https://github.com/linq2db/linq2db/pull/1447 regression

Comment (MaceWindu) [reply Source/LinqToDB/DataProvider/PostgreSQL/PostgreSQLDataProvider.cs]: fixed

**Pull request merged**